src/test/unit/encoders/SimHashDocumentEncoderTest.cpp

```diff
@@ -0,0 +1,237 @@
+    encoder1.encode(testDoc1, output1);
+    encoder2.encode(testDoc2, output2);
+    encoder3.encode(testDoc3, output3);
+    encoder4.encode(testDoc4, output4);
+
+    ASSERT_GT(output3.getOverlap(output4), output2.getOverlap(output3));
+    ASSERT_GT(output2.getOverlap(output3), output1.getOverlap(output3));
+    ASSERT_GT(output1.getOverlap(output3), output1.getOverlap(output4));
+  }
+
+  // Test encoding a simple corpus with 'tokenSimilarity' Off (default). Tokens
+  // of similar spelling will NOT affect the output in shared manner, but apart.
+  TEST(SimHashDocumentEncoder, testTokenSimilarityOff) {
+    SimHashDocumentEncoderParameters params;
+    params.size = 400u;
+    params.sparsity = 0.33f;
+    params.tokenSimilarity = false;
+
+    SDR output1({ params.size });
+    SDR output2({ params.size });
+    SDR output3({ params.size });
+    SDR output4({ params.size });
+    SimHashDocumentEncoder encoder1(params);
+    SimHashDocumentEncoder encoder2(params);
+    SimHashDocumentEncoder encoder3(params);
+    SimHashDocumentEncoder encoder4(params);
+    encoder1.encode(testDoc1, output1);
+    encoder2.encode(testDoc2, output2);
+    encoder3.encode(testDoc3, output3);
+    encoder4.encode(testDoc4, output4);
+
+    ASSERT_GT(output1.getOverlap(output2), output2.getOverlap(output3));
+    ASSERT_GT(output2.getOverlap(output3), output3.getOverlap(output4));
+    ASSERT_GT(output3.getOverlap(output4), output1.getOverlap(output3));
+  }
+
+  // Test encoding with weighted tokens. Make sure output changes accordingly.
+  TEST(SimHashDocumentEncoder, testTokenWeightMap) {
+    SimHashDocumentEncoderParameters params;
+    params.size = 400u;
+    params.sparsity = 0.33f;
+    params.tokenSimilarity = false;
+
+    SimHashDocumentEncoder encoderA(params);
+    SDR outputA({ params.size });
+    encoderA.encode(testDocMap1, outputA);
+
+    SimHashDocumentEncoder encoderB(params);
+    SDR outputB({ params.size });
+    encoderB.encode(testDocMap2, outputB);
+
+    SimHashDocumentEncoder encoderC(params);
+    SDR outputC({ params.size });
+    encoderC.encode(testDocMap3, outputC);
+
+    ASSERT_GT(outputA.getOverlap(outputC), outputA.getOverlap(outputB));
+    ASSERT_GT(outputA.getOverlap(outputB), outputB.getOverlap(outputC));
+  }
+
+  // Test encoding unicode text with 'tokenSimilarity' on
+  TEST(SimHashDocumentEncoder, testUnicodeSimilarityOn) {
+    SimHashDocumentEncoderParameters params;
+    params.size = 400u;
+    params.sparsity = 0.33f;
+    params.tokenSimilarity = true;
+
+    SimHashDocumentEncoder encoderA(params);
+    SDR outputA({ params.size });
+    encoderA.encode(testDocUni1, outputA);
+
+    SimHashDocumentEncoder encoderB(params);
+    SDR outputB({ params.size });
+    encoderB.encode(testDocUni2, outputB);
+
+    ASSERT_GT(outputA.getOverlap(outputB), 65u);
+  }
+
+  // Test encoding unicode text with 'tokenSimilarity' Off
+  TEST(SimHashDocumentEncoder, testUnicodeSimilarityOff) {
```

Comment:
Can you please add some real-world example (documents) of how this encoder is used, would be useful? Ie how do you intend to use the encoder.

Reply:
sure thing: https://github.com/htm-community/htm.core/pull/603/commits/c376950fea93c79313ed367aa8c786674ff5fb6f